grade-scripts/trackers-from-requests.js

```diff
@@ -0,0 +1,158 @@
+
+    execSync(`mkdir -p ${outputPath}`)
+
+    // get initial file data
+    let siteDataArray = fileManager.getSiteData(inputPath, outputPath, fileForSubset, true)
+
+    for (let siteData of siteDataArray) {
+        let url = siteData.url
+        let hostname = url.replace(/https?:\/\//, '')
+
+        let trackersBlocked = {}
+        let trackersNotBlocked = {}
+        let totalBlocked = 0
+
+        // requests are stored as a tuple like: [url, requestType]
+        siteData.requests.forEach((request) => {
+            let tracker = trackers.isTracker(request[0], url, request[1])
+
+            if (tracker) {
+                if (tracker.block) {
+                    if (!trackersBlocked[tracker.parentCompany]) {
+                        trackersBlocked[tracker.parentCompany] = {}
+                    }
+                    trackersBlocked[tracker.parentCompany][tracker.url] = tracker
+                    totalBlocked += 1
+                } else {
+                    if (!trackersNotBlocked[tracker.parentCompany]) {
+                        trackersNotBlocked[tracker.parentCompany] = {}
+                    }
+                    trackersNotBlocked[tracker.parentCompany][tracker.url] = tracker
+                }
+            }
+        })
+
+        console.log(chalk.green(`${url}: ${totalBlocked} trackers`))
+
+        let outputData = {
+            url: siteData.url,
+            trackersBlocked,
+            trackersNotBlocked,
+            totalBlocked
+        }
+
+        if (siteData.rank) {
+            outputData.rank = siteData.rank
+        }
+
+        fs.writeFileSync(`${outputPath}/${hostname}.json`, JSON.stringify(outputData))
+    }
+}
+
+const calculateTrackerPrevalence = () => {
+    console.log('calculating tracker prevalence data...')
+
+    // get all the files that we've just processed
+    let siteDataArray = fileManager.getSiteData(outputPath)
+    let entityList = listManager.getList('entityList')
+
+    let networksSeen = {}
+
+    // count up all the tracker networks seen on each site
+    siteDataArray.forEach((siteData) => {
+        let networksOnThisSite = {}
+
+        // get all the tracker networks seen on a specific site
+        Object.keys(siteData.trackersBlocked).forEach((network) => {
+            Object.keys(siteData.trackersBlocked[network]).forEach((trackerUrl) => {
+                let parent = entityList[trackerUrl] || network
+                networksOnThisSite[parent] = true
+            })
+        })
+        Object.keys(siteData.trackersNotBlocked).forEach((network) => {
+            Object.keys(siteData.trackersNotBlocked[network]).forEach((trackerUrl) => {
+                let parent = entityList[trackerUrl] || network
+                networksOnThisSite[parent] = true
+            })
+        })
+
+        // add them to the global counts
+        Object.keys(networksOnThisSite).forEach((network) => {
+            if (network === 'unknown') return
```

Comment:
Note, this is a place where I deviated from the original prototype, which counted "unknown" as a single tracker network. Prevalence % might be different.

(Also, I've only got prevalence %'s for 15k sites while developing, we'll want to rerun them before integrating with the extension.)

Reply:
Note: we still want this information - so we can tell how we need to expand our entity map.